File: 프로그래머스/3/43163. 단어 변환/단어 변환.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int solution(string begin, string target, string[] words) {
        
        var set = new HashSet<string>(words);
        if(!set.Contains(target)) return 0;
        
        bool[] visited = new bool[words.Length];
        var q = new Queue<(string word, int depth)>();
        q.Enqueue((begin, 0));
        
        while(q.Count > 0)
        {
            var (cur, depth) = q.Dequeue();
            if(cur == target) return depth;
            
            for(int i = 0; i < words.Length; i++)
            {
                if(!visited[i] && IsAdjacent(cur, words[i]))
                {
                    q.Enqueue((words[i], depth + 1));
                }
            }
        }
        return 0;
    }
    private bool IsAdjacent(string a, string b)
    {
        int diff = 0;
        for(int i = 0; i < a.Length; i++)
        {
            if(a[i] != b[i]) diff++;
            if(diff > 1) return false;
        }
        return diff == 1;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int answer;

bool check(string a, string b)
{
    int diff = 0;
    for(int i = 0; i < a.size(); i++)
    {
        if(a[i] != b[i]) diff++;
    }
    return diff == 1;
}

int solution(string begin, string target, vector<string> words) {
    queue<pair<string, int>> q;
    bool visited[51] = {0};
    
    q.push({begin, 0});
    
    while(q.size())
    {
        string now = q.front().first;
        int depth = q.front().second;
        q.pop();
        
        if(now == target)
            return depth;
        
        for(int i = 0; i < words.size(); i++)
        {
            if(visited[i]) continue;
            if(check(now, words[i]))
            {
                visited[i] = true;
                q.push({words[i], depth + 1});
            }
        }
    }
    return 0;
}